runtime/compiler/control/JITServerHelpers.cpp

```diff
@@ -37,63 +37,88 @@ bool         JITServerHelpers::_serverAvailable = true;
-   return (name->length + sig->length + (2 * sizeof(U_16)));
+   return OMR::alignNoCheck(str->length, sizeof(str->length)) + sizeof(str->length);
+   }
+
+static size_t
+getPackedUTF8Size(const J9UTF8 *str, const J9ROMClass *romClass)
+   {
+   return JITServerHelpers::isAddressInROMClass(str, romClass) ? 0 : getUTF8Size(str);
+   }
+
+// If the string points outside of the contiguous part of origRomClass, appends
+// it (with padding) at curPos and updates the SRP to the string in the packed
+// ROMClass. Returns the new value of curPos.
+static uint8_t *
+packUTF8(uint8_t *curPos, const J9UTF8 *origStr, const J9ROMClass *origRomClass, J9SRP &srp)
+   {
+   if (JITServerHelpers::isAddressInROMClass(origStr, origRomClass))
+      return curPos;
+
+   size_t size = getUTF8Size(origStr);
+   memcpy(curPos, origStr, size);
+   static_assert(sizeof(origStr->length) == 2, "UTF8 length field is not 2 bytes large");
+   if (!OMR::alignedNoCheck(origStr->length, sizeof(origStr->length)))
+      curPos[size - 1] = '\0';
+   NNSRP_SET(srp, curPos);
+   return curPos + size;
    }
 
 // Packs a ROMClass into a std::string to be transferred to the server.
-// The name and signature of all methods are appended to the end of the cloned class body and the
-// self referential pointers to them are updated to deal with possible interning. The method names
-// and signature are needed on the server but may be interned globally on the client.
+// Some of the name and signature strings are interned and stored outside
+// of the ROMClass body. Such strings are appended to the end of the cloned
+// ROMClass body and the self referential pointers to them are updated.
 static std::string
 packROMClass(J9ROMClass *origRomClass, TR_Memory *trMemory)
    {
+   size_t totalSize = origRomClass->romSize;
    J9UTF8 *className = J9ROMCLASS_CLASSNAME(origRomClass);
-   size_t classNameSize = className->length + sizeof(U_16);
+   totalSize += getPackedUTF8Size(className, origRomClass);
 
    J9ROMMethod *romMethod = J9ROMCLASS_ROMMETHODS(origRomClass);
-   size_t totalSize = origRomClass->romSize + classNameSize;
    for (size_t i = 0; i < origRomClass->romMethodCount; ++i)
       {
-      totalSize += methodStringsLength(romMethod);
+      totalSize += getPackedUTF8Size(J9ROMMETHOD_NAME(romMethod), origRomClass);
+      totalSize += getPackedUTF8Size(J9ROMMETHOD_SIGNATURE(romMethod), origRomClass);
       romMethod = nextROMMethod(romMethod);
       }
 
+   // Check if the contiguous part of the ROMClass already contains all the strings visited above
+   if (totalSize == origRomClass->romSize)
+      return std::string((char *)origRomClass, origRomClass->romSize);
+
    J9ROMClass *romClass = (J9ROMClass *)trMemory->allocateHeapMemory(totalSize);
    if (!romClass)
       throw std::bad_alloc();
    memcpy(romClass, origRomClass, origRomClass->romSize);
+   // Update the size to include the strings that will be appended at the end
+   romClass->romSize = totalSize;
```

Comment:
We didn't do this before. Was this a bug?

Reply:
Yes, but it didn't actually lead to any problems since these strings are only accessed directly and without checking that they are within the ROMClass bounds.